crates/ruff/src/jupyter/notebook.rs

```diff
@@ -130,60 +168,142 @@ impl JupyterNotebook {
             )));
         }
 
-        Ok(notebook)
-    }
-
-    /// Concatenates all cells into a single virtual file and builds an index that maps the content
-    /// to notebook cell locations
-    pub fn index(&self) -> (String, JupyterIndex) {
-        let mut jupyter_index = JupyterIndex {
-            // Enter a line number (1-based), get back the cell (1-based)
-            // 0 index is just padding
-            row_to_cell: vec![0],
-            // Enter a line number (1-based), get back the row number in the cell (1-based)
-            // 0 index is just padding
-            row_to_row_in_cell: vec![0],
-        };
-        let size_hint = self
+        let valid_code_cells = notebook
             .cells
             .iter()
-            .filter(|cell| cell.cell_type == CellType::Code)
-            .count();
+            .enumerate()
+            .filter(|(_, cell)| is_valid_code_cell(cell))
+            .map(|(pos, _)| pos)
+            .collect::<Vec<_>>();
```

Comment:
This is somewhat expensive because it requires copying over all code blocks. Would a "virtual" notebook work where we only store the indices of the code cells and than skip all non-code cells while iterating?

Reply:
I believe that's what is happening here unless I'm mistaken with what you're referring to as "expensive"?

Here, we're collecting all the indices which are valid code cells, storing it on `Notebook` to be used wherever required.

Reply:
You're right, lol. I'm sorry